crates/ra_hir_ty/src/_match.rs

```diff
@@ -0,0 +1,1174 @@
+                // to perform specialization to see that our wildcard will cover
+                // the `Some(false)` case.
+                let mut constructor = None;
+                for pat in matrix.heads() {
+                    if let Some(c) = pat_constructor(cx, pat)? {
+                        constructor = Some(c);
+                        break;
+                    }
+                }
+
+                if let Some(constructor) = constructor {
+                    if let Constructor::Enum(e) = constructor {
+                        // For enums we handle each variant as a distinct constructor, so
+                        // here we create a constructor for each variant and then check
+                        // usefulness after specializing for that constructor.
+                        let mut found_unimplemented = false;
+                        for constructor in
+                            cx.db.enum_data(e.parent).variants.iter().map(|(local_id, _)| {
+                                Constructor::Enum(EnumVariantId { parent: e.parent, local_id })
+                            })
+                        {
+                            let matrix = matrix.specialize_constructor(&cx, &constructor)?;
+                            let v = v.expand_wildcard(&cx, &constructor)?;
+
+                            match is_useful(&cx, &matrix, &v) {
+                                Ok(Usefulness::Useful) => return Ok(Usefulness::Useful),
+                                Ok(Usefulness::NotUseful) => continue,
+                                _ => found_unimplemented = true,
+                            };
+                        }
+
+                        if found_unimplemented {
+                            Err(MatchCheckNotImplemented)
+                        } else {
+                            Ok(Usefulness::NotUseful)
+                        }
+                    } else {
+                        let matrix = matrix.specialize_constructor(&cx, &constructor)?;
+                        let v = v.expand_wildcard(&cx, &constructor)?;
+
+                        is_useful(&cx, &matrix, &v)
+                    }
+                } else {
+                    Ok(Usefulness::NotUseful)
+                }
+            }
+            _ => {
+                // Either not all constructors are covered, or the only other arms
+                // are wildcards. Either way, this pattern is useful if it is useful
+                // when compared to those arms with wildcards.
+                let matrix = matrix.specialize_wildcard(&cx);
+                let v = v.to_tail();
+
+                is_useful(&cx, &matrix, &v)
+            }
+        }
+    }
+}
+
+#[derive(Debug)]
+/// Similar to TypeCtor, but includes additional information about the specific
+/// value being instantiated. For example, TypeCtor::Bool doesn't contain the
+/// boolean value.
+enum Constructor {
+    Bool(bool),
+    Tuple { arity: usize },
```

Comment:
rustc represents this with `Single`, as far as I understand. So it contains a bit less information, but the only place you really need the arity is when expanding the pattern -- I assume rustc uses the type information there :thinking: we can leave it like this for now, but I think a simpler representation (for this whole enum) would be better here.

Reply:
> the only place you really need the arity is when expanding the pattern -- I assume rustc uses the type information there

I believe you are correct here. It may have just been my being newer to this codebase that I preferred converting things into this `Constructor` type. 

> we can leave it like this for now, but I think a simpler representation (for this whole enum) would be better here.

I'll try to keep this under consideration in follow up MRs as I implement the rest of this match checking functionality.